File: benchmark/benchmark.py
'''
benchmark utility comparing numpy and ndarray, with the ha-ndarray
'''
# std
import time

# 3rd
import numpy as np
from matplotlib import pyplot as plt

# local
from ha_ndarray import ndarray_test
from ha_ndarray import ha_ndarray_test


if __name__ == "__main__":

    data_types = ['uint8', 'uint16', 'uint32', 'uint64', 'float32', 'float64']
    data_types_2 = ['u8', 'u16', 'u32', 'u64', 'f32', 'f64']
    sizes = [2**1, 2**3, 2**5, 2**7, 2**9, 2**11]#, 2**13]#, 2**15]
    operation = ['add', 'sub', 'mul', 'div', 'dot']

    # dictionary to times
    timings_py = {
        'add': np.zeros((len(data_types), len(sizes))),
        'sub': np.zeros((len(data_types), len(sizes))),
        'mul': np.zeros((len(data_types), len(sizes))),
        'div': np.zeros((len(data_types), len(sizes))),
        'dot': np.zeros((len(data_types), len(sizes)))
    }

    for op in operation:  # Skip 'add' as it's already done
        for k in range(len(data_types)):
            for j in range(len(sizes)):
                results = []
                print(f'Py - type: {data_types[k]}, size: {sizes[j]}, operation: {op}')
                for replicates in range((sizes[-1] // sizes[j]) + 2):
                    a = np.random.rand(sizes[j], sizes[j]).astype(data_types[k])
                    b = np.random.rand(sizes[j], sizes[j]).astype(data_types[k])
                    b[b == 0] = 1
                    start = time.perf_counter()
                    if op == 'sub':
                        c = a - b
                    elif op == 'mul':
                        c = a * b
                    elif op == 'div':
                        c = a / b
                    elif op == 'dot':
                        c = np.dot(a, b)
                    elif op == 'add':
                        c = a + b
                    else: # throw shade
                        raise ValueError(f'op: {op} not supported')
                    results.append(time.perf_counter() - start)

                timings_py[op][k, j] = np.array(results).mean()


    timings_ndarray = ndarray_test()
    timings_ndarray = {k: np.array(v) for k, v in timings_ndarray.items()}

    # i havent learnt how to use your crate yet
    timings_ha_array = ha_ndarray_test()
    timings_ha_array = {k: np.array(v) for k, v in timings_ha_array.items()}

    # Plotting results for each operation
    print(f'len data_types: {len(data_types)}, len operation: {len(operation)}')
    fig, axes = plt.subplots(len(data_types), len(operation), figsize=(15, 10))

    lines = ['-', '--', '-.', ':', '-']

    sizes_sq = [sq ** 2 for sq in sizes]

    for i, _ in enumerate(data_types):
        for j, op in enumerate(operation):
            if len(data_types) == 1 and len(operation) == 1:
                ax = axes
            elif len(operation) == 1:
                ax = axes[i]
            else:
                ax = axes[i, j]
            ax.set_xlabel('size (# elements 2D)')
            ax.set_ylabel('time (s)')
            ax.set_title(f'op: {op}; datatype: {data_types[i]}')
            ax.set_xscale('log')
            ax.set_yscale('log')
            ax.grid(True)

            # plotting
            ax.plot(sizes_sq, timings_py[op][i, :], label=f'np', color=[1, 0, 1])
            ax.plot(sizes_sq, timings_ndarray[op][i, :], label=f'nd', color=[1, 0.5, 0])
            ax.plot(sizes_sq, timings_ha_array[op][i, :], label=f'ha', color=[0, 0.5, 1.0])

    axes[i, j].legend()

    plt.tight_layout()
    plt.savefig('fig.png')
    plt.show()

File: benchmark/benchmark.rs
mod benchmark_ha;
mod benchmark_nd;
mod benchmark_py;
use std::collections::HashMap;

fn print_table(
    data_types: &[&str],
    operations: &[&str],
    sizes: &[usize],
    timings_py: &HashMap<&str, Vec<Vec<f64>>>,
    timings_nd: &HashMap<&str, Vec<Vec<f64>>>,
    timings_ha: &HashMap<&str, Vec<Vec<f64>>>,
) {
    println!(
        "\n\n\n|------------|-----------|------|-----------|-----------|-----------|-------|-------|"
    );
    println!(
        "| {:<10} | {:<9} | {:<4} | {:<9} | {:<9} | {:<9} | {:<5} | {:<5} |",
        "Operation", "Data Type", "Size", "Py Timing", "ND Timing", "HA Timing", "Py/HA", "ND/HA"
    );
    println!(
        "|------------|-----------|------|-----------|-----------|-----------|-------|-------|"
    );

    for &op in operations {
        for &dtype in data_types {
            for &size in sizes {
                let py_time = timings_py[op][dtype_index(dtype)][size_index(size, sizes)];
                let nd_time = timings_nd[op][dtype_index(dtype)][size_index(size, sizes)];
                let ha_time = timings_ha[op][dtype_index(dtype)][size_index(size, sizes)];

                let py_ha = py_time / ha_time;
                let nd_ha = nd_time / ha_time;

                println!(
                    "| {:<10} | {:<9} | {:<4} | {:<9.7} | {:<9.7} | {:<9.7} | {:<5.2} | {:<5.2} |",
                    op, dtype, size, py_time, nd_time, ha_time, py_ha, nd_ha
                );
            }
        }
    }

    println!(
        "|------------|-----------|------|-----------|-----------|-----------|-------|-------|"
    );
}

fn dtype_index(dtype: &str) -> usize {
    match dtype {
        "uint8" => 0,
        "uint16" => 1,
        "uint32" => 2,
        "uint64" => 3,
        "float32" => 4,
        "float64" => 5,
        _ => unreachable!(),
    }
}

fn size_index(size: usize, sizes: &[usize]) -> usize {
    sizes.iter().position(|&s| s == size).unwrap()
}

fn main() {
    println!("Hi");

    let size_max: u32 = 9;

    let data_types: Vec<&str> = vec!["uint8", "uint16", "uint32", "uint64", "float32", "float64"];
    let operations: Vec<&str> = vec!["add", "sub", "mul", "div", "dot"];
    let sizes: Vec<usize> = (1..=size_max).step_by(2).map(|x| 2_usize.pow(x)).collect();

    let timings_py: HashMap<&str, Vec<Vec<f64>>> =
        benchmark_py::py_ndarray_test(data_types.clone(), operations.clone(), sizes.clone());

    let timings_nd: HashMap<&str, Vec<Vec<f64>>> =
        benchmark_nd::ndarray_test(data_types.clone(), operations.clone(), sizes.clone());

    let timings_ha: HashMap<&str, Vec<Vec<f64>>> =
        benchmark_ha::ha_ndarray_test(data_types.clone(), operations.clone(), sizes.clone());

    print_table(
        &data_types,
        &operations,
        &sizes,
        &timings_py,
        &timings_nd,
        &timings_ha,
    );
}
